docs/proposals/sync-namespace-ondemand.md

```diff
@@ -0,0 +1,60 @@
+
+## Proposal
+
+### Introduce a new SyncNamespace API that would facilitate on-demand synchronization of a namespace
+
+The SyncNamespace API would perform the following actions:
+
+1. The Ruler that receives the call determines which Rulers need to be notified to sync/reload the namespace. This can be accomplished by:
+   1. Querying the ring for the Rulers that can service the user’s namespaces. For instance, when shuffle sharding is utilized, calling ring.ShuffleShard(userID, shardSize), would return the sub-ring for the user
+   2. Listing all the rule groups for that namespace by calling ListRuleGroupsForUserAndNamespace
+   3. Further filtering the Rulers it needs to notify by asking the sub-ring for Rulers for each rule group from the previous step
+2. The Ruler notifies all the other Rulers from the previous step to sync/reload the namespace
+3. The Rulers that were notified to sync/reload the namespace would:
+   1. List all the rule groups for that namespace by calling ListRuleGroupsForUserAndNamespace
+   2. Filter for the rule groups it owns
+   3. Loads the rule groups from storage by invoking `LoadRuleGroups`
+   4. Synchronize the rule groups to Prometheus. While syncing a specific namespace, the Ruler can read the rest of the namespace files from local file system and update Prometheus with a combined list of all the namespace files. This is necessary to prevent Prometheus from stopping/removing rule groups from namespaces that are not being updated. A change will also need to be made to prevent Cortex from removing other users from the Ruler
+   5. As an alternative to previous step, we could propose a change to Prometheus to enable updating of a single namespace. At present, Prometheus manager in rules package has an `Update` method that not only refreshes the rule groups but also removes rule groups not present in the specified list of files. This is why Cortex must provide all the files for a user to Prometheus. Introduction of a method to update a single namespace without affecting other namespaces would be advantageous
+
+![SyncNamespace On-Demand](/website/static/images/proposals/sync-namespace-on-demand.png)
+
+### API
+
+POST `/api/v1/rules/{namespace}/sync`
+
+### Scenarios:
+
+#### Competing SyncNamespace Calls:
+
+When a single Ruler receives multiple SyncNamespace calls, the Ruler will block while updating the local files and synchronizing with Prometheus. This should prevent multiple threads from potentially corrupting rule files
+
+
+#### Simultaneous SyncNamespace and Periodic Sync:
+
+A new buffered channel will be created on the Ruler and the existing syncRules function has a for/select clause. A new case statement will be added to the existing select clause. This will prevent SyncNamespace and period sync from running concurrently
```

Comment:
Since we prevent multiple concurrent syncs to avoid corrupting files, what happens if we have high volume of `sync` requests? With a new buffered channel this will like cause a delay in the execution of the period sync right? If that's that case then should there be a mechanism where the period sync has priority and when it runs it will clear the new buffered channels because the per namespace syncs will not be necessary since period sync will update all namespaces.

Reply:
Unlike the periodic sync, the on-demand sync will only download files from external storage for a particular namespace which should be significantly smaller number of files and therefore processing should be quicker. In addition, on-demand sync will delegate to different set of Rulers based on user/rule group name/namespace name. Also, once the on-demand sync is caught up, there should not be any more namespaces that require synchronization

Reply:
I agree, my point here is that if we use a buffered channel to prevent period sync and SyncNamespace from running at the same time, won't a consistent medium/high tps request to to the sync endpoint cause the channel to be almost always be full preventing periodic sync to get a change to execute?